stonesoup/hypothesiser/distance.py

```diff
@@ -87,3 +91,65 @@ def hypothesise(self, track, detections, timestamp, **kwargs):
                         measurement_prediction))
 
         return MultipleHypothesis(sorted(hypotheses, reverse=True))
+
+
+class MultiDistanceHypothesiser(Hypothesiser):
+    """Prediction Hypothesiser based on a set of Measures
+
+    Similar to the :class:`~.DistanceHypothesiser` this class scores prediction-detection pairs
+    using their distance in a measurement space. However, this class attempts to use the
+    detection's measurement model to determine a suitable Measure and distance to use. These are
+    provided by the user.
+    To do so, the class utilises a list of :class:`~.DistanceHypothesiser` and corresponding
+    measurement models. When a detection is received of particular model, the class will attempt
+    to find the model's corresponding hypothesiser and use its output.
+    """
+
+    hypothesisers: Sequence[DistanceHypothesiser] = Property(
+        doc="Sequence of hypothesisers to utilise output of.")
+    measurement_models: Sequence[MeasurementModel] = Property(
+        doc="Sequence of measurement models that will be compared against incoming detections. "
+            "Must be same length as hypothesisers sequence."
+    )
+
+    def __init__(self, *args, **kwargs):
+
+        super().__init__(*args, **kwargs)
+
+        if len(self.hypothesisers) != len(self.measurement_models):
+            raise ValueError("Number of hypothesisers must equal number of measurement models in "
+                             "MultiDistanceHypothesiser")
+
+    def hypothesise(self, track, detections, timestamp, **kwargs):
+        """Passes detections that do not have measurement model or have a model with no
+        corresponding hypothesiser, to the first hypothesiser in the sequence. Uses the
+        null-hypothesis of the first hypothesiser and will ignore those of the other
+        :attr:`hypothesisers`."""
+
+        # Keep track of detections attributed to each hypothesiser
+        hypothesisers_detections = {hypothesiser: set() for hypothesiser in self.hypothesisers}
+
+        # True detection hypotheses
+        for detection in detections:
+            try:
+                detection_measurement_model = detection.measurement_model
+                index_of_model = self.measurement_models.index(detection_measurement_model)
```

Comment:
I'm not sure this will work where you have models which are more dynamic e.g. with changing translation or rotation offsets.
It would be better as an instance check? Maybe it could use MRO to find nearest match?

Reply:
Good point. Perhaps limiting the user to only defining measurement model type, rather than instance would be better?